### DISCUSSION TITLE
DEFAULT CURRENT_TIMESTAMP datetime error for PHP < 5.6

### DISCUSSION THREAD

**Discussion by apmuthu** [timschofield/webERP#175] [Problems / Bugs]
[Source](https://stackoverflow.com/questions/23153822/mysql-datetime-default-current-timestamp-error)

The `DEFAULT CURRENT_TIMESTAMP` support for a `DATETIME` (datatype) was added in MySQL 5.6.

In 5.5 and earlier versions, this applied only to `TIMESTAMP` (datatype) columns.

It is possible to use a `BEFORE INSERT` trigger in 5.5 to assign a default value to a column.
````
 DELIMITER $$

 CREATE TRIGGER ...
 BEFORE INSERT ON mytable
 FOR EACH ROW
 BEGIN
    IF NEW.mycol IS NULL THEN
       SET NEW.mycol = NOW();
    END IF;
 END$$
````
The import of the `stockserialitems` table fails as it has a `DATETIME` field with the said defaults.

**Comment by TurboPT** [comment]
I have not (yet) checked all the tables, but is this issue **only** with the stockserialitems table?

While I have no personal issues with triggers, know that there are (currently) no other triggers in the system, so it _might_ be easy enough (or better?) to change the data type for greater version coverage, if we're only considering one table.

**Comment by apmuthu** [comment]
Just changing the fieldtype to be `TIMESTAMP` for now should be sufficient for backwards compatibility - The year 2037 is quite sometime away before the Y2K scenario for Linux 32 bit kicks in. There is no other case of `DATETIME` field type having a `DEFAULT CURRENT_TIMESTAMP` in the sqls for now. Otherwise, we need to remove the `DEFAULT CURRENT_TIMESTAMP` and make sure that in all places in the code where a record insertion for this table occurs it is properly computed and provided for.

**Comment by apmuthu** [comment]
The following can be committed for backwards compatibility:
````
--- sql/country_sql/default.sql	Sat Mar 31 06:09:13 2018
+++ sql/country_sql/default.sql	Sat Mar 31 12:40:58 2018
@@ -2920,7 +2920,7 @@
   `expirationdate` datetime NOT NULL DEFAULT '0000-00-00 00:00:00',
   `quantity` double NOT NULL DEFAULT '0',
   `qualitytext` text NOT NULL,
-  `createdate` DATETIME NULL DEFAULT CURRENT_TIMESTAMP,
+  `createdate` timestamp NULL DEFAULT CURRENT_TIMESTAMP,
   PRIMARY KEY (`stockid`,`serialno`,`loccode`),
   KEY `StockID` (`stockid`),
   KEY `LocCode` (`loccode`),
--- sql/country_sql/demo.sql	Sat Mar 31 06:09:13 2018
+++ sql/country_sql/demo.sql	Sat Mar 31 12:41:11 2018
@@ -2922,7 +2922,7 @@
   `expirationdate` datetime NOT NULL DEFAULT '0000-00-00 00:00:00',
   `quantity` double NOT NULL DEFAULT '0',
   `qualitytext` text NOT NULL,
-  `createdate` DATETIME NULL DEFAULT CURRENT_TIMESTAMP,
+  `createdate` timestamp NULL DEFAULT CURRENT_TIMESTAMP,
   PRIMARY KEY (`stockid`,`serialno`,`loccode`),
   KEY `StockID` (`stockid`),
   KEY `LocCode` (`loccode`),
--- sql/upgrade4.14.1-4.14.2.sql	Sat Mar 31 06:09:13 2018
+++ sql/upgrade4.14.1-4.14.2.sql	Sat Mar 31 12:42:14 2018
@@ -31,7 +31,7 @@
 ALTER TABLE pcashdetails MODIFY receipt text COMMENT 'Column redundant. Replaced by receipt file upload. Nov 2017.';
 INSERT INTO `scripts` (`script` ,`pagesecurity` ,`description`) VALUES ('BankAccountBalances.php',  '1',  'Shows bank accounts authorised for with balances');
 
-ALTER TABLE `stockserialitems` ADD `createdate` DATETIME NULL DEFAULT CURRENT_TIMESTAMP, ADD INDEX ( `createdate` );
+ALTER TABLE `stockserialitems` ADD `createdate` timestamp NULL DEFAULT CURRENT_TIMESTAMP, ADD INDEX ( `createdate` );
 UPDATE stockserialitems SET createdate = NULL;
 
 UPDATE stockserialitems as stockserialitems SET createdate = 
````
**Note** that the `stockserialitems.createdate` field is not present in the `weberpchina.sql` file.

[sql_timespamp.diff.txt](https://github.com/timschofield/webERP-svn/files/1865256/sql_timespamp.diff.txt)

**Comment by apmuthu** [comment]
Ignore above diff.
Better fix in #28.

**Comment by timschofield** [comment]
Can this issue be closed now?

**Comment by apmuthu** [comment]
Yes. Working on some fixes to the sqls too.
